### ISSUE TITLE
Provide page-specific security headers via .withHeaders(...)

### ISSUE THREAD

**Issue by bwbecker** [playframework/playframework#6011]
I'm using the tinyMCE html editor in one of my pages.  It breaks the Content Security Policies that I have set for the rest of my site.

As a workaround, I tried returning

```scala
Ok(EditNoteView(uwid, noteid, filledForm))
  .withHeaders(CONTENT_SECURITY_POLICY_HEADER → "default-src 'self' 'unsafe-inline'")
```

in my controller, thinking that would override the headers set in the SecurityHeadersFilter.  No such luck.  The filter replaces any CSP headers already set.

I'd like to rewrite the ```headers(req: RequestHeaders):Seq[(String, String)]``` method in SecurityHeadersFilter (see line 123 in https://github.com/playframework/playframework/blob/2.5.x/framework/src/play-filters-helpers/src/main/scala/play/filters/headers/SecurityHeadersFilter.scala) so that it does not replace existing headers in the result.

Questions:

- Does the community think this is the right thing to do?
- Should page-specific security header overrides via ```.withHeaders``` require enabling via the config file?
- I believe this requires changing the ```headers``` signature to include the result.  One option is ```headers(request:RequestHeaders, result:Result)```;  another option is ```headers(request:RequestHeaders, result:Option[Result]=None)```.  The first option is a breaking change but the code is cleaner.  The second option leaves us with messier code for the long term.  Which would you prefer?

Byron

**Comment by bwbecker** [comment]
This was first discussed in Play's google group. Greg Methvin responded as follows:

Hi Byron,

In the filter the headers are applied to the result after your action executes, so it would override anything that already exists. I think maybe it should avoid setting the headers if they're already set, to allow the use case you're talking about. Perhaps whether it does that should be configurable as well. If you want you could create an issue on github for this.

If you're writing a custom filter, you could still reuse the SecurityHeadersConfig (so configuration still works normally) and simply bind your own SecurityHeadersFilter that does what you want.

We'd welcome pull requests to improve the filter as well.

Regards,
Greg

**Comment by mkurz** [comment]
[That's the line](https://github.com/playframework/playframework/blob/2.5.1/framework/src/play-filters-helpers/src/main/scala/play/filters/headers/SecurityHeadersFilter.scala#L136) which needs to changed to avoid setting a header that has been set before already.

**Comment by bwbecker** [comment]
@mkurz Except the only header I need to change is CONTENT_SECURITY_POLICY_HEADER.  I'd like to leave the other 4 in place.  So I think the best approach is to modify the headers helper so that for each of the five security headers it defaults to values in the Result and if there isn't one, uses the value in the config.  In that case the line you point to can remain as is except for the call to headers.

**Comment by mkurz** [comment]
@bwbecker You are welcome to provide a pull request.

**Comment by gmethvin** [comment]
The workaround would be to create a custom `SecurityHeadersFilter` that looks like this:
```scala
class MySecurityHeadersFilter @Inject() (config: SecurityHeadersConfig) extends SecurityHeadersFilter(config) {
  override protected def headers(request: RequestHeader) = {
    super.headers(request).filter { case (name, _) => request.headers.get(name).isEmpty }
  }
}
```
then just add that to your filters class.

Assuming you're using the default Guice setup, you shouldn't need to enable/disable any modules, since the `SecurityHeadersConfig` should already be bound and Guice can automatically create the filter from the constructor.

**Comment by gmethvin** [comment]
We can easily add a fix for this in master. It's possible some users rely on the current behavior, so we'd probably want to add a setting in config, e.g. `overrideHeaders = true`.

Since the workaround is fairly straightforward and the fix would require a few workarounds for binary compatibility, I'm not sure it'd be worth the trouble to backport to 2.5.

**Comment by bwbecker** [comment]
Greg -- I don't think your solution gets it.  You need to check against the headers in the result, not the request.  I think it should be:

```scala
class CustomSecurityHeadersFilter @Inject()(config: SecurityHeadersConfig)
  extends SecurityHeadersFilter(config) {

  protected def headers(request: RequestHeader, result: Result): Seq[(String, String)] = {
    super.headers(request).filter { case (name, _) ⇒ result.header.headers.get(name).isEmpty }
  }

  override def apply(next: EssentialAction) = EssentialAction { req =>
    import play.api.libs.iteratee.Execution.Implicits.trampoline
    next(req).map { result ⇒
      result.withHeaders(headers(req, result): _*)
    }
  }
}
```

Therefore, I think the change to master needs to include the signature change to headers.

I agree that maintaining binary compatibility in 2.5.x isn't worth it and that the override in config is a good idea.

If you agree with the fix, above, would you like me to work on a pull request (change the code, test suite, documentation)?  Anything else?

**Comment by gmethvin** [comment]
Yes, you're right, it's the result headers. Not sure what I was thinking there.If you'd like to submit a PR that's great.

**Comment by otves** [comment]
@gmethvin Which version will contain the fix this bug? Play 2.5.12 doesn't contain fix.

**Comment by jtal** [comment]
Does anyone use the "Embedded router" method for javascript routes as described below?

It creates inline javascript which, in chrome, raises an error about the Content Security Policy rule. Should this method be deprecated now?

https://www.playframework.com/documentation/2.5.x/ScalaJavascriptRouting
